rfcs/20201201-cpp-gradients.md

```diff
@@ -0,0 +1,415 @@
+
+
+#### GradientRegistry
+
+We provide a registry to store the mapping from op type to factory functions that return the `GradientFunction` for an op’s instance. The factory function takes as input the `ForwardOperation`, which contains metadata from the forward operation, and returns a `GradientFunction`. This allows gradient function authors to control which inputs/outputs of the forward op to keep around by increasing the ref-count on `AbstractTensorHandle`.
+
+Additionally, we provide a utility function `RegisterNotDifferentiable` to mark an op as non-differentiable. This can be used to implement `tf.no_gradient`. We also provide a `NotDifferentiableGradientFunction` which returns nullptr output gradients. This can be used to implement `tf.stop_gradient`.
+
+
+```
+struct ForwardOperation {
+ public:
+  string op_name;
+  std::vector<AbstractTensorHandle*> inputs;
+  std::vector<AbstractTensorHandle*> outputs;
+  std::vector<int64> skip_input_indices;
+  AttrBuilder attrs;
+};
+
+using GradientFunctionFactory = std::function<GradientFunction*(const ForwardOperation& op)>;
+
+// Map from op name to a GradientFunctionFactory.
+class GradientRegistry {
+ public:
+  Status Register(const string& op,
+                  GradientFunctionFactory gradient_function_factory);
+  Status Lookup(const ForwardOperation& op,
+                std::unique_ptr<GradientFunction>* gradient_function) const;
+
+ private:
+  absl::flat_hash_map<string, GradientFunctionFactory> registry_;
+};
+
+class NotDifferentiableGradientFunction : public GradientFunction {
+  Status Compute(
+      AbstractContext* ctx,
+      absl::Span<AbstractTensorHandle* const> grad_outputs,
+      absl::Span<AbstractTensorHandle*> grad_inputs) override;
+};
+Status RegisterNotDifferentiable(GradientRegistry* registry, const string& op);
+```
+
+
+Examples:
+
+
+```
+GradientFunction* AddRegisterer(const ForwardOperation& op) {
+  return new AddGradientFunction;
+}
+
+GradientFunction* ExpRegisterer(const ForwardOperation& op) {
+  return new ExpGradientFunction(op.outputs[0]);
+}
+```
+
+
+
+#### Tape
+
+The API for C++ `Tape` is very similar to python’s `tf.GradientTape`. The implementation for this interface is almost entirely shared with the C++ tape in `c/eager/tape.h`. 
+
+
+```
+class Tape {
+ public:
+  explicit Tape::Tape(bool is_persistent);
+
+  // Adds this tensor to the list of watched tensors.
+  //
+  // This is a no-op if the tensor is already being watched either from an
+  // earlier call to GradientTape::Watch or being an output of an op with
+  // watched inputs.
+  void Watch(const AbstractTensorHandle* tensor);
+
+  // Records an operation with given inputs and outputs
+  // on the tape and marks all its outputs as watched if at
+  // least one input of the op is watched and has a trainable dtype.
+  // op_name is optional and is used for debugging only.
+  void RecordOperation(absl::Span<AbstractTensorHandle* const> inputs,
```

Comment:
In python our execution path automatically calls `recordoperation` as needed. When users are looking to have C++ only training pipelines, are they looking to manually define their model in terms of raw C++ ops? And if so, would they have to manually call `RecordOperation` for each op?

Reply:
Ah, I see you mention this in the `#### Automatic variable tracking` section below.

Reply:
That's a great question! I am glad you brought this up. Another part of the C++ API that I am working on, and that is briefly mentioned in this doc, are the multi-backend [op execution interfaces](https://cs.opensource.google/search?q=f:abstract_(context%7Coperation%7Ctensor_handle).h&ss=tensorflow%2Ftensorflow:tensorflow%2F) for implementing this earlier [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20191203-single-eager-graph-path.md). I imagine there being a tape execution context which just traces ops on the tape and delegates to some backing context for say tracing or eager execution or another `TapeContext`. This is implemented [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/gradients/tape/). 

I wrote this up but didn't add it to this doc because the execution interfaces are still not finalized yet and may change with @allenlavoie's work on op handlers. I am happy to add more details on [`TapeContext`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/gradients/tape/tape_context.h), [`TapeOperation`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/gradients/tape/tape_operation.h) here if that would be useful.

Automatic variable tracking is separate. It will remain a higher level feature for now.

Reply:
I think it makes sense to clarify this a bit in the doc. Let me add something.

Reply:
Added.